src/acl/Address.cc

```diff
@@ -18,3 +22,42 @@ Acl::Address::~Address()
     delete next;
 }
 
+std::optional<Ip::Address>
+Acl::Address::findAddressCandidate(HttpRequest * const request) const
+{
+    if (std::holds_alternative<MatchClientTcpDst>(addressSource)) {
+        if (request && request->clientConnectionManager.valid()) {
+            auto localAddr = request->clientConnectionManager->clientConnection->local;
+            localAddr.port(0);
+            return localAddr;
+        }
+
+        // not DBG_IMPORTANT or Assure() because the client may simply be gone by now
+        debugs(17, 3, "WARNING: Skipping tcp_outgoing_address match_client_tcp_dst rule");
+        return std::nullopt;
+    }
+
+    return std::get<Ip::Address>(addressSource);
+}
+
+std::ostream &
+Acl::operator <<(std::ostream &os, const Address::AddressSource &source)
+{
+    if (std::holds_alternative<Address::MatchClientTcpDst>(source)) {
+        os << "match_client_tcp_dst";
+        return os;
+    }
+
+    const auto &addr = std::get<Ip::Address>(source);
+
+    if (addr.isAnyAddr()) {
+        // XXX: Use squid.conf syntax (e.g., "any_addr") for all special values
+        os << "autoselect";
```

Comment:
We should fix this XXX, but doing so is outside this PR scope. See commit 6fb1483 message for more details.

Reply:
Moving this to a STL stream was a good call. The code is so much nicer 👍